File: leetCode/23-Merge-k-Sorted-Lists.py
#too lazy to write it in c
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
        
class Solution:        
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        list1 = []
        for i in range(0, len(lists)):
            temp = lists[i]
            while temp:
                list1.append(temp.val)
                temp = temp.next
        
        if not list1:
            return None
        
        list1.sort()
        newList = ListNode(val=list1[0])
        temp = newList
        for i in range(1, len(list1)):
            temp.next = ListNode(val=list1[i])
            temp = temp.next
        
        
        return newList
      
     

File: leetCode/23-Merge-k-Sorted-Lists.c
//c implementation
 struct ListNode {
     int val;
     struct ListNode *next;
 };



struct ListNode *mergeLists(struct ListNode *a, struct ListNode *b) {
    
    
    struct ListNode *result = malloc(sizeof(struct ListNode));
    result->next = NULL;
    struct ListNode *temp = result;
    
    while(a || b) {
        if(a && b) {
            if(a->val > b->val) {
                temp->next = b;
                b = b->next;
            } else if(a->val < b->val) {
                temp->next = a;
                a = a->next;
            } else if(a->val == b->val) {
                temp->next = a;
                a = a->next;
                temp->next->next = b;
                b = b->next;
                temp = temp->next;
            }
            
            temp = temp->next;
        } else {
            temp->next = a != NULL ? a : b; 
            break;
        } 
    }
    temp = result->next;
    free(result);
    return temp;   
}

struct ListNode* mergeKLists(struct ListNode **lists, int listsSize){
    struct ListNode *merged;
    
    if(listsSize == 0) {
        return NULL;
    } else {
        merged = lists[0];
        for(int i = 1; i < listsSize; ++i) {
            merged = mergeLists(merged, lists[i]);
        }
    }

    return merged;
}
